independent-projects/tools/devtools-common/src/main/java/io/quarkus/devtools/commands/handlers/ListCategoriesCommandHandler.java

```diff
@@ -0,0 +1,96 @@
+package io.quarkus.devtools.commands.handlers;
+
+import io.quarkus.devtools.commands.ListCategories;
+import io.quarkus.devtools.commands.data.QuarkusCommandException;
+import io.quarkus.devtools.commands.data.QuarkusCommandInvocation;
+import io.quarkus.devtools.commands.data.QuarkusCommandOutcome;
+import io.quarkus.devtools.messagewriter.MessageWriter;
+import io.quarkus.devtools.project.BuildTool;
+import io.quarkus.registry.catalog.Category;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.function.BiConsumer;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ * Instances of this class are thread-safe. It lists extension categories according to the options passed in as properties of
+ * {@link QuarkusCommandInvocation}
+ */
+public class ListCategoriesCommandHandler implements QuarkusCommandHandler {
+
+    private static final String ID_FORMAT = "%-20s";
+    private static final String FULL_FORMAT = "%-30s %-20s %s";
+
+    private static final String LIST_EXTENSIONS_HINT = "\nTo list extensions in given category, use:\n%s";
+    private static final String MORE_INFO_HINT = "\nTo get more information, append `%s` to your command line.";
+
+    @Override
+    public QuarkusCommandOutcome execute(QuarkusCommandInvocation invocation) throws QuarkusCommandException {
+
+        final MessageWriter log = invocation.log();
+        final boolean cli = invocation.getValue(ListCategories.FROM_CLI, false);
+        final boolean batchMode = invocation.getValue(ListCategories.BATCH_MODE, false);
+        final String format = invocation.getValue(ListCategories.FORMAT, "");
+
+        final Collection<Category> categories = invocation.getExtensionsCatalog().getCategories();
+
+        if (!cli) {
+            log.info("Available Quarkus extension categories: ");
+        }
+
+        BiConsumer<MessageWriter, Category> formatter;
+        switch (format.toLowerCase()) {
+            case "full":
+                log.info(String.format(FULL_FORMAT, "Category", "Id", "Description"));
+                formatter = this::fullFormatter;
+                break;
+            case "id":
+            default:
+                formatter = this::idFormatter;
+                break;
+        }
+
+        categories.stream()
+                .sorted(Comparator.comparing(Category::getName))
+                .forEach(c -> formatter.accept(log, c));
+
+        final BuildTool buildTool = invocation.getQuarkusProject().getBuildTool();
+        boolean isGradle = BuildTool.GRADLE.equals(buildTool) || BuildTool.GRADLE_KOTLIN_DSL.equals(buildTool);
+
+        if (!cli) {
+            if (StringUtils.isBlank(format)) {
+                if (isGradle) {
+                    log.info(MORE_INFO_HINT, "--format=full");
+                } else {
+                    log.info(MORE_INFO_HINT, "-Dquarkus.extension.format=full");
+                }
+            }
+
+            if (isGradle) {
+                log.info(LIST_EXTENSIONS_HINT,
```

Comment:
This is the wrong place to log this information, isn't it? 
The path to the executable isn't known here.. and this doesn't take jbang into account at all. 

The cli handles emitting this information (and will, via its runners): and I can see that all of this is on the !cli path... 

The Maven/Gradle plugins should be tracing this information on their own (based on what they're reading to set this handler up). We shouldn't be reconstructing the original command here.

You can still define the constants for the hint, though I would put them on the ListCategoriesCommand (rather than on the handler).

Reply:
> The cli handles emitting this information (and will, via its runners): and I can see that all of this is on the !cli path...

Why shouldn't this be on `!cli` path? The line you marked will print "To list extensions... use `./gradlew listExtensions ...`", so that seems correct. On the `cli == true` path it will print `quarkus extensions list...`.

I'm not sure about your meaning here, we are not guessing the original command (not the full command), but we show how a user can list _extensions_ based on the list of _categories_ that was just displayed, with the same tool he just used.

We already do the very same thing in the `ListExtensionsCommandHandler`, which shows hint with a command to add an extension based on the list of extensions that was printed, snippet:

```
        if (!cli) {
            // ...
            if (isGradle) {
                log.info("\nAdd an extension to your project by adding the dependency to your " +
                        "build.gradle or use `./gradlew addExtension --extensions=\"artifactId\"`");
            } else {
                log.info("\nAdd an extension to your project by adding the dependency to your " +
                        "pom.xml or use `./mvnw quarkus:add-extension -Dextensions=\"artifactId\"`");
            }
        }
```

Yes, it's guessing about the `./mvnw` / `./gradlew` commands, but those are reasonable guesses.

I added the same hint for the `cli==true` case, which uses `quarkus` as a command name. That should probably be improved, I'm not sure if we officially want to use `quarkus` or `qs` but there is probably a way to detect what's used in reality.